src/lib/url.js

```diff
@@ -0,0 +1,64 @@
+const queryParams = ['include', 'fields', 'sort'];
+
+export const parseInclude = include => {
+  return include ? include.split(',') : [];
+};
+
+export const parseFields = query => {
+  let fields = {};
+
+  if (query) {
+    for (let [key, value] of query) {
+      if (key.startsWith('fields') && key.indexOf('[') > -1) {
+        const type = key.substring(key.indexOf('[') + 1, key.indexOf(']'));
+        fields[type] = new Set(value.split(','));
+      }
+    }
+  }
+  return fields;
+};
+
+export const parseJsonApiUrl = fromUrl => {
+  const url = new URL(fromUrl);
+  const query = url.searchParams;
+
+  return {
+    protocol: url.protocol,
+    host: url.host,
+    port: url.port,
+    path: url.pathname,
+    query: {
+      filter: query.get('filter'),
+      include: parseInclude(query.get('include')),
+      fields: parseFields(query.entries()),
+      sort: query.get('sort') || [],
```

Comment:
`sort` has the same syntax as include. Maybe we can `s/parseInclude/parseListParameter` and `s/parseFields/parseDictionaryParameter` and after that, `parseDictionaryParameter` could reuse `parseListParameter` for each instance of a `field[*]` query parameter, since that is a list too.

Reply:
I like that a lot, using a more generic name. I hadn't gotten to sort yet but it seemed like it had some more complexity, such as the "normal" syntax and with sort order.

Reply:
Ugh, good call. I totally forgot about the "normal" thing on https://www.drupal.org/docs/8/modules/jsonapi/sorting.

I don't know why we support that non-spec'd way when the spec seems perfectly sufficient. It's probably my fault and I've just forgotten that it is lol; maybe a leftover from https://www.drupal.org/project/entityqueryapi